packages/api/server/ws-client.mts

```diff
@@ -214,7 +214,7 @@ export default class WebSocketServer {
     if (event === 'subscribe') {
       conn.subscriptions.push(topic);
       conn.reply(topic, 'subscribed', { id: payload.id });
-      channel.onJoinCallback(topic, conn.socket);
+      channel.onJoinCallback(topic, conn);
```

Comment:
The special `onJoinCallback` was implemented as a quick hack before #332. However, now that #332 is in place, we should make the callback signature the same as all other handlers:

```diff
+type HandlerType = (
+  payload: Record<string, any>,
+  context: MessageContextType,
+  conn: ConnectionContextType,
+) => void;
+
 /**
  * Channel is responsible for dispatching incoming messages for a given topic.
  *
@@ -52,15 +58,11 @@ export class Channel {
     string,
     {
       schema: z.ZodTypeAny;
-      handler: (
-        payload: Record<string, any>,
-        context: MessageContextType,
-        conn: ConnectionContextType,
-      ) => void;
+      handler: HandlerType;
     }
   > = {};

-  onJoinCallback: (topic: string, ws: WebSocket) => void = () => {};
+  onJoinCallback: HandlerType = () => {};

   constructor(topic: string) {
     this.topic = topic;
@@ -130,7 +132,7 @@ export class Channel {
     return this;
   }

-  onJoin(callback: (topic: string, ws: WebSocket) => void) {
+  onJoin(callback: HandlerType) {
     this.onJoinCallback = callback;
     return this;
   }
@@ -214,7 +216,11 @@ export default class WebSocketServer {
     if (event === 'subscribe') {
       conn.subscriptions.push(topic);
       conn.reply(topic, 'subscribed', { id: payload.id });
-      channel.onJoinCallback(topic, conn.socket);
+      channel.onJoinCallback(
+        payload,
+        { topic: match.topic, event: event, params: match.params },
+        conn,
+      );
       return;
     }
```

The code in server/channels/app.mts can then do:

```typescript
.onJoin((_payload, context, conn) => {
  const appExternalId = (context as AppContextType).params.appId;
  // ...
});
```